R/read-policy.R

```diff
@@ -48,3 +48,83 @@ check_keys <- function(policy, verbose = FALSE){
   
   policy
 }
+
+#' Get the Category Names for Policy
+#'
+#' This function identifies the names of all categories in a policy list.
+#'
+#' @param policy R list of a valid policy file
+#' @param verbose Whether or not to return an alert of categories
+#' 
+#' @examples
+#' get_categories(policy_demo, verbose = TRUE)
+#' 
+#' @return A vector of the names of the categories in the policy R list
+#' @importFrom purrr map
+#' @importFrom cli cli_alert_info cli_div cli_text cli_end
+#' @export
+
+get_categories <- function(policy, verbose = FALSE){
+  policy <- flatten_policy(policy)
+  
+  categories <- purrr::map(policy$categories, "category") |>
+    unlist()
```

Comment:
map_chr() can only iterate over a length-1 vector, not a list

Reply:
map_chr() can iterate over a list per the documentation and I use the cousin function map_lgl in the Canvas PR to iterate over a list perfectly fine. The requirement is that the returned value for each element in the iterable must be length 1. That is the case here since each category name should be just a single string (ie a length 1 char vector).

Reply:
Still doesn't work though because it returns a vector of assignments (i.e. not a vector of length)
Try this code below, and you'll see what I mean...
category <- list(
  category = "Overall Grade",
  aggregation = "weighted_mean",
  assignments = list(
    list(
      category = "Labs",
      aggregation = "min_score",
      assignments = list(
        list(
          category = "Lab 1",
          aggregation = "max_score",
          assignments = c("Lab 1.1", "Lab 1.2")
        ),
        list(
          category = "Lab 2",
          aggregation = "max_score",
          assignments = c("Lab 2.1", "Lab 2.2", "Lab 2.3")
        )
      )
    ),
    list(
      category = "Quiz",
      aggregation = "weighted_by_points",
      assignments = list(
        list(
          category = "Quiz 1",
          aggregation = "max_score",
          assignments = c("Quiz 1 Individual", "Quiz 1 Group")
        ),
        list(
          category = "Quiz 2",
          aggregation = "max_score",
          assignments = c("Quiz 2 Individual", "Quiz 2 Group")
        )
      )
    )
  )
)
if (!is.list(category$assignments)){
  return(category$assignments)
} else{
  purrr::map_chr(category$assignments, function(nested_cat){
    get_nested_assignments(nested_cat)
  })
}

Reply:
Yeah map_chr won't work for use in get_nested_assignments() for this reason (which is why I did not recommend it). I believe for the use in get_categories() it should work however. Unless I'm mistaken, the "category" attribute should always be length 1?

Reply:
Oh, gotcha, just changed it